mitmproxy/addons/maplocal.py

```diff
@@ -0,0 +1,135 @@
+    """Join a Path element with an untrusted str.
+
+    This is a convenience wrapper for werkzeug's safe_join,
+    raising a ValueError if the path is malformed."""
+    untrusted_parts = Path(untrusted).parts
+    joined = safe_join(
+        root.as_posix(),
+        *untrusted_parts
+    )
+    if joined is None:
+        raise ValueError("Untrusted paths.")
+    return Path(joined)
+
+
+def file_candidates(url: str, spec: MapLocalSpec) -> typing.List[Path]:
+    """
+    Get all potential file candidates given a URL and a mapping spec ordered by preference.
+    This function already assumes that the spec regex matches the URL.
+    """
+    m = re.search(spec.regex, url)
+    assert m
+    if m.groups():
+        suffix = m.group(1)
+    else:
+        suffix = re.split(spec.regex, url, maxsplit=1)[1]
+        suffix = suffix.split("?")[0]  # remove query string
+
+    suffix = re.sub(r"[^0-9a-zA-Z\-_.=(),/]", "_", suffix.strip("/"))
+
+    if suffix:
+        try:
+            return [
+                _safe_path_join(spec.local_path, suffix),
+                _safe_path_join(spec.local_path, f"{suffix}/index.html")
+            ]
+        except ValueError:
+            return []
+    else:
+        return [spec.local_path / "index.html"]
+
+
+class MapLocal:
+    def __init__(self):
+        self.replacements: typing.List[MapLocalSpec] = []
+
+    def load(self, loader):
+        loader.add_option(
+            "map_local", typing.Sequence[str], [],
+            """
+            Map remote resources to a local file using a pattern of the form
+            "[/flow-filter]/url-regex/file-or-directory-path", where the
+            separator can be any character.
+            """
+        )
+
+    def configure(self, updated):
+        if "map_local" in updated:
+            self.replacements = []
+            for option in ctx.options.map_local:
+                try:
+                    spec = parse_map_local_spec(option)
+                except ValueError as e:
+                    raise exceptions.OptionsError(f"Cannot parse map_local option {option}: {e}") from e
+
+                self.replacements.append(spec)
+
+    def request(self, flow: http.HTTPFlow) -> None:
+        if flow.reply and flow.reply.has_message:
+            return
+
+        url = flow.request.pretty_url
+
+        any_spec_matches = False
+        for spec in self.replacements:
+            if spec.matches(flow) and re.search(spec.regex, url):
+                any_spec_matches = True
+
+                local_file: typing.Optional[Path] = None
+
+                if spec.local_path.is_file():
+                    local_file = spec.local_path
+                elif spec.local_path.is_dir():
+                    for candidate in file_candidates(url, spec):
+                        if candidate.is_file():
+                            local_file = candidate
+                            break
+
+                headers = {}
```

Comment:
How about we add a `Server: mitmproxy/MapLocal` (or whatever) header? When looking at the response in dev tools or similar it would make it super clear what's happening. Makes it a no-brainer to check if the mapping is correctly working.

Reply:
Good point, we have [`version.MITMPROXY`](https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/version.py#L7) for this.

Reply:
:+1:, done